test/unit/org/apache/cassandra/db/filter/ColumnFilterTest.java

```diff
@@ -36,143 +48,388 @@
+        testRoundTrips(filter);
+        assertEquals("[v2[1:3]]", filter.toString());
+        assertFetchedQueried(true, true, filter, v2);
+        assertFetchedQueried(false, false, filter, v1, s1, s2);
+        assertCellFetchedQueried(true, true, filter, v2, path1, path2, path3);
+        assertCellFetchedQueried(false, false, filter, v2, path0, path4);
+        assertCellFetchedQueried(false, false, filter, s2, path0, path1, path2, path3, path4);
+    }
+
+    @Test
+    public void testSelectCellSliceFromStatic() throws Exception
+    {
+        ColumnFilter filter = ColumnFilter.selectionBuilder().slice(s2, path1, path3).build();
+        testRoundTrips(filter);
+        assertEquals("[s2[1:3]]", filter.toString());
+        assertFetchedQueried(true, true, filter, s2);
+        assertFetchedQueried(false, false, filter, v1, v2, s1);
+        assertCellFetchedQueried(false, false, filter, v2, path0, path1, path2, path3, path4);
+        assertCellFetchedQueried(true, true, filter, s2, path1, path2, path3);
+        assertCellFetchedQueried(false, false, filter, s2, path0, path4);
+    }
+
+    @Test
+    public void testSelectColumnsWithCellsAndSlices() throws Exception
+    {
+        ColumnFilter filter = ColumnFilter.selectionBuilder().add(v1).add(s1).slice(v2, path0, path2).select(v2, path4).select(s2, path0).slice(s2, path2, path4).build();
+        testRoundTrips(filter);
+        assertEquals("[s1, s2[0], s2[2:4], v1, v2[0:2], v2[4]]", filter.toString());
+        assertFetchedQueried(true, true, filter, v1, v2, s1, s2);
+        assertCellFetchedQueried(true, true, filter, v2, path0, path1, path2, path4);
+        assertCellFetchedQueried(false, false, filter, v2, path3);
+        assertCellFetchedQueried(true, true, filter, s2, path0, path2, path3, path4);
+        assertCellFetchedQueried(false, false, filter, s2, path1);
+    }
+
+    // select with metadata
+
+    @Test
+    public void testSelectSimpleColumnWithMetadata() throws Exception
+    {
+        Consumer<ColumnFilter> check = filter -> {
+            testRoundTrips(filter);
+            assertFetchedQueried(true, true, filter, v1);
+            if (anyNodeOn30)
+            {
+                assertEquals("*/*", filter.toString());
+                assertFetchedQueried(true, true, filter, s1, s2, v2);
+                assertCellFetchedQueried(true, true, filter, v2, path0, path1, path2, path3, path4);
+                assertCellFetchedQueried(true, true, filter, s2, path0, path1, path2, path3, path4);
+            }
+            else
+            {
+                assertEquals("*/[v1]", filter.toString());
+                assertFetchedQueried(true, false, filter, s1, s2, v2);
+                assertCellFetchedQueried(true, false, filter, v2, path0, path1, path2, path3, path4);
+                assertCellFetchedQueried(true, false, filter, s2, path0, path1, path2, path3, path4);
+            }
```

Comment:
Nit: maybe it would be a bit easier to spot the differences between mixed and normal modes this way:
```suggestion
            assertEquals(anyNodeOn30 ? "*/*" : "*/[v1]", filter.toString());
            assertFetchedQueried(true, anyNodeOn30, filter, s1, s2, v2);
            assertCellFetchedQueried(true, anyNodeOn30, filter, v2, path0, path1, path2, path3, path4);
            assertCellFetchedQueried(true, anyNodeOn30, filter, s2, path0, path1, path2, path3, path4);
```

Reply:
Actually I'd not do that. Especially because to keep easy merge path from 3.0 -> 3.11 -> trunk. Note that in trunk we have 3 scenarios and playing with `? / :` would no longer make it more readable. 

Now we can easily see that the assertions for 3.0 in cassandra-3.0, cassandra-3.11 and trunk are identical. Same for assertions for 3.11 in cassandra-3.11 and trunk.